---
title: "RNA-Seq analysis"
output:
  html_document: 
    toc: true
  pdf_document: 
    toc: true
  word_document: 
    toc: true
params:
  out_dir: NA
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = F)
```

\pagebreak

# Summary

```{r comment="", results='asis'}
library(knitr)
out_dir=params$out_dir
gen_sum=read.csv(paste0(out_dir,"/reports/general_report.csv"),sep="\t",header = F,row.names = 1)
kable(gen_sum,col.names = "Summary table")
sam_sum=read.csv(paste0(out_dir,"/reports/sample_report.csv"),sep="\t",header = T)
kable(sam_sum)
```

\pagebreak

# FastQ quality

```{r comment="", results='asis', fig.align='center'}
library(magick)
library(rlist)
sam=as.character(sam_sum$sample_name)
for (i in 1:length(sam)) {
  cat(paste0("\n\n##",sam[i]),"\n")
  img1=image_read(list.files(path=paste0(params$out_dir,"/Quality/",sam[i]),pattern="_screen.png$",full.names = T)[1])
  plot(img1)
  cat("\n\n")
  img2=image_read(list.files(path=paste0(params$out_dir,"/Quality/",sam[i]),pattern="_screen.png$",full.names = T)[2])
  plot(img2)
  cat("\n\n")
  img3=image_read(paste0(list.dirs(path=paste0(params$out_dir,"/Quality/",sam[i]),recursive = F)[1],"/Images/per_base_quality.png"))
  img3=image_annotate(img3,list.dirs(path=paste0(params$out_dir,"/Quality/",sam[i]),recursive = F,full.names = F)[1],size=30)
  plot(img3)
  cat("\n\n")
  img4=image_read(paste0(list.dirs(path=paste0(params$out_dir,"/Quality/",sam[i]),recursive = F)[2],"/Images/per_base_quality.png"))
  img4=image_annotate(img4,list.dirs(path=paste0(params$out_dir,"/Quality/",sam[i]),recursive = F,full.names = F)[2],size=30)
  plot(img4)
  cat("\n\n")
}
```

\pagebreak

# BAM quality

```{r comment="", fig.align='center', out.width = "90%", dpi = 500}

processFile = function(filepath,title) {
  cat(title,"\n")
  con = file(filepath, "r")
  while ( TRUE ) {
    line = readLines(con, n = 1)
    if ( length(line) == 0 ) {
      break
    }
    cat(line,"\n")
  }
  close(con)
} 

if (gen_sum[3,1]=="hisat") {dir_a="/HISAT2/"}
if (gen_sum[3,1]=="tophat") {dir_a="/TopHat2/"}
for (i in 1:length(sam)) {
  cat(paste0("\n\n",sam[i]),"\n\n")
  stat=list.files(path=paste0(out_dir,dir_a,sam[i],"/RSeQC"),patter="bam_stat.txt",full.names = T)
  processFile(stat,paste0("#### Statistics ",sam[i], " ####"))
  read_d=list.files(path=paste0(out_dir,dir_a,sam[i],"/RSeQC"),patter="read_distribution.txt",full.names = T)
  processFile(read_d, paste0("\n\n#### Read distribution ",sam[i]," ####\n"))
  cat("\n\n")
  img_inn=image_read_pdf(list.files(path=paste0(out_dir,dir_a,sam[i],"/RSeQC"),patter="inner_distance_plot.pdf",full.names = T))
  img_inn=image_annotate(img_inn, paste0("Inner distance plot ",sam[i]),gravity = "North",size = 70)
  plot(img_inn)
}

```

\pagebreak

# Differential expression analysis

```{r fig.align='center', out.width = "90%", dpi = 500}
pca=image_read(list.files(path=paste0(out_dir,"/Quantification_and_DEA"),
                          pattern = "-pca.png",
                          full.names = T))
print(pca)
cat("\n\n")
volcano=image_read(list.files(path=paste0(out_dir,"/Quantification_and_DEA"),
                              pattern = "-volcanoplot.png",
                              full.names = T))
print(volcano)
cat("\n\n")
heat1=image_read(list.files(path=paste0(out_dir,"/Quantification_and_DEA"),
                            pattern = "-heatmap-topVarGenes.png",
                            full.names = T))
heat1=image_annotate(heat1,"Top variance genes heatmap",gravity = "North",size = 30)
print(heat1)
cat("\n\n")
heat2=image_read(list.files(path=paste0(out_dir,"/Quantification_and_DEA"),
                            pattern = "-heatmap-samples.png",
                            full.names = T))
print(heat2)

```

\pagebreak

# Meta-analysis

## Gene ontology

```{r fig.align='center', out.width = "100%", dpi = 500}
tree=image_read_pdf(paste0(out_dir,"/Meta-analysis/Gene_ontology/treemap_GO_fc1.5_pv0.05.pdf"))
print(tree)
cat("\n\n")
dot_go=image_read_pdf(paste0(out_dir,"/Meta-analysis/Gene_ontology/dotplot_GO.pdf"))
print(dot_go[1])
cat("\n\n")
print(dot_go[2])
cat("\n\n")
print(dot_go[3])
cat("\n\n")
cnet_go=image_read_pdf(paste0(out_dir,"/Meta-analysis/Gene_ontology/cnetplot_GO.pdf"))
print(cnet_go[1])
cat("\n\n")
print(cnet_go[2])
cat("\n\n")
print(cnet_go[3])
cat("\n\n")
```

## Pathway analysis

```{r fig.align='center', out.width = "100%", dpi = 500}
dot=image_read_pdf(paste0(out_dir,"/Meta-analysis/Pathway_analysis/dotplot_pathways.pdf"))
print(dot)
cat("\n\n")
cnet_path=image_read_pdf(paste0(out_dir,"/Meta-analysis/Pathway_analysis/cnetplot_pathways.pdf"))
print(cnet_path)
```

\pagebreak

# References

* S. Andrews, “FastQC: A quality control tool for high throughput sequence data.,” Http://Www.Bioinformatics.Babraham.Ac.Uk/Projects/Fastqc/, 2010. 
* "S. Andrews, “FastQ Screen,” http://www.bioinformatics.babraham.ac.uk/projects/fastq_screen/, 2011." 
* H. Li and R. Durbin, “Fast and accurate short read alignment with Burrows-Wheeler transform,” Bioinformatics, vol. 25, no. 14, pp. 1754–1760, 2009.
* H. Li, “A statistical framework for SNP calling, mutation discovery, association mapping and population genetical parameter estimation from sequencing data,” Bioinformatics, vol. 27, no. 21, pp. 2987–2993, 2011.
* D. Kim, G. Pertea, C. Trapnell, H. Pimentel, R. Kelley, and S. L. Salzberg, “TopHat2: Accurate alignment of transcriptomes in the presence of insertions, deletions and gene fusions,” Genome Biol., vol. 14, no. 4, 2013.
* L. Wang, S. Wang, and W. Li, “RSeQC: quality control of RNA-seq experiments,” Bioinformatics, vol. 28, no. 16, pp. 2184–2185, 2012.
* D. Kim, B. Langmead, and S. L. Salzberg, “HISAT: A fast spliced aligner with low memory requirements,” Nat. Methods, vol. 12, no. 4, pp. 357–360, 2015.
* Y. Liao, G. K. Smyth, and W. Shi, “FeatureCounts: An efficient general purpose program for assigning sequence reads to genomic features,” Bioinformatics, vol. 30, no. 7, pp. 923–930, 2014.
* C. Trapnell, B. A. Williams, G. Pertea, A. Mortazavi, G. Kwan, M. J. Van Baren, S. L. Salzberg, B. J. Wold, and L. Pachter, “Transcript assembly and quantification by RNA-Seq reveals unannotated transcripts and isoform switching during cell differentiation,” Nat. Biotechnol., vol. 28, no. 5, pp. 511–515, 2010.
* L. A. Goff, C. Trapnell, and D. Kelley, “cummeRbund: Analysis, exploration, manipulation, and visualization of Cufflinks high-throughput sequencing data,” R Packag. Version 2.2, 2012.
* M. D. Robinson, D. J. McCarthy, and G. K. Smyth, “edgeR: a Bioconductor package for differential expression analysis of digital gene expression data.,” Bioinformatics, vol. 26, no. 1, pp. 139–40, 2010.
* D. J. McCarthy, Y. Chen, and G. K. Smyth, “Differential expression analysis of multifactor RNA-Seq experiments with respect to biological variation,” Nucleic Acids Res., vol. 40, no. 10, pp. 4288–4297, 2012.
* M. I. Love, W. Huber, et al., “Moderated estimation of fold change and dispersion for RNA-seq data with DESeq2,” Genome Biol., vol. 15, no. 12, p. 550, 2014.
* Yuan Tang, Masaaki Horikoshi, and Wenxuan Li. “ggfortify: Unified Interface to Visualize Statistical Result of Popular R Packages.” The R Journal 8.2, pp. 478-489, 2016.
* Masaaki Horikoshi and Yuan Tang, “ggfortify: Data Visualization Tools for Statistical Analysis Results.”, 2016.
* J. Graffelman, “calibrate: Calibration of Scatterplot and Biplot Axes.,” R Packag. version 1.7.2., 2013.
* R. Gentleman, V. Carey, W. Huber, and F. Hahne, “genefilter: methods for filtering genes from high-throughput experiments.,” R Packag. version 1.60.0., 2017.
* G. R. Warnes, B. Bolker, L. Bonebakker, R. Gentleman, W. H. A. Liaw, T. Lumley, M. Maechler, A. Magnusson, S. Moeller, M. Schwartz, and B. Venables, “gplots: Various R Programming Tools for Plotting Data,” R Packag. version 3.0.1., 2016.
* Yihui Xie, “DT: A Wrapper of the JavaScript Library 'DataTables'.” R package version 0.4., 2018.
* Jeroen Ooms, “magick: Advanced Graphics and Image-Processing in R.” R package version 1.9., 2018.
* Almende B.V., Benoit Thieurmel and Titouan Robert, “visNetwork: Network Visualization using 'vis.js' Library.” R package version 2.0.4., 2018.
* Winston Chang, Joe Cheng, JJ Allaire, Yihui Xie and Jonathan McPherson, “shiny: Web Application Framework for R.” R package version 1.0.5., 2017.
